### DISCUSSION TITLE
Long press buttons event

### DISCUSSION THREAD

**Discussion by wormstein** [EdNekebno/DGTCentaurMods#51] [Show and tell]
Hey,
I am running out of ideas on this... Let me know if you spot what's wrong.
So, looks like each button have two serial calls on push and rekease, Base on this I am trying to implement long press detection in boardfunction.eventsThread()

The while loop should check for 3 second if the release event executes meaning this is a short press. For some reason, the while loop executes just once if I keep the button pressed. but if I press it and release immediately it will execute the if condition inside the while...I even see the serial call, wich is good. But holding the button for 3 seconds is not returning BTNPLAYLONG meaning the while loop never breaks as it should...
It's a very strange behaviour.

I wanted to start with BTNPLAY so I am having this code. Ignore lots of debug lines...

``` python
              if (resp.hex() == "b10010065000140a0504000000002a68"):
                    print("BTNPLAY Button pressed")
                    print(resp)
                    print("-----------------")
                    timeout = time.time() + 3
                    while time.time() < timeout:
                        #print("Entering loop...")
                        #btnup = "b10011065000140a0500040000017d7a37"
                        #print(btnup)
                        #tosend = bytearray(b'\x94\x06\x50\x6a')
                        ser.write(tosend)
                        expect = bytearray(b'\xb1\x00\x06\x06\x50\x0d')
                        #print(tosend)
                        resp = ser.read(10000)
                        resp = bytearray(resp)
                        print("[<-]"+str(resp))
                        print(resp[9])
                        if (bytearray(resp) != expect):
                            if ( resp[9] == 0 ):
                                print("BTNPLAY Sent")
                                buttonPress = BTNPLAY   # PLAY
                                break
                    print("END OF WHILE")
                    print("BTNPLAYLONG Sent")
                    buttonPress = BTNPLAYLONG
```

**Comment by EdNekebno** [comment]
Interesting. Nice discovery that there is a press and release event. It's a good idea. My brain isn't working - but I don't think you're actually checking btnup and after the while loop you're always setting it to BTNPLAYLONG regardless. Also you may be overwhelming the board with the amount of communication. They data seems to be buffered so you can afford to sleep a bit in the loop.

I'll run a quick test.

**Comment by EdNekebno** [comment]
This is a really good spot. I make all the relevant strings (in hex).

UP
Press: 0xb1 0x0 0x11 0x6 0x50 0x0 0x14 0xa 0x5 0x8 0x0 0x0 0x0 0x0 0x7d 0x3c 0x7c
Release: 0xb1 0x0 0x10 0x6 0x50 0x0 0x14 0xa 0x5 0x0 0x8 0x0 0x0 0x1 0x12 0x55
DOWN
Press: 0xb1 0x0 0x10 0x6 0x50 0x0 0x14 0xa 0x5 0x2 0x0 0x0 0x0 0x0 0x61 0x1d
Release: 0xb1 0x0 0x11 0x6 0x50 0x0 0x14 0xa 0x5 0x0 0x2 0x0 0x0 0x1 0x7d 0xe 0x49
BACK
Press: 0xb1 0x0 0x11 0x6 0x50 0x0 0x14 0xa 0x5 0x1 0x0 0x0 0x0 0x0 0x7d 0x47 0x0
Release: 0xb1 0x0 0x11 0x6 0x50 0x0 0x14 0xa 0x5 0x0 0x1 0x0 0x0 0x1 0x7d 0x34 0x6e
TICK
Press: 0xb1 0x0 0x11 0x6 0x50 0x0 0x14 0xa 0x5 0x10 0x0 0x0 0x0 0x0 0x7d 0x17 0x5f
Release: 0xb1 0x0 0x11 0x6 0x50 0x0 0x14 0xa 0x5 0x0 0x10 0x0 0x0 0x1 0x7d 0x45 0xe
QUESTION
Press: 0xb1 0x0 0x10 0x6 0x50 0x0 0x14 0xa 0x5 0x40 0x0 0x0 0x0 0x0 0x6d 0x67
Release: 0xb1 0x0 0x10 0x6 0x50 0x0 0x14 0xa 0x5 0x0 0x40 0x0 0x0 0x1 0x39 0x34
PLAY
Press: 0xb1 0x0 0x10 0x6 0x50 0x0 0x14 0xa 0x5 0x4 0x0 0x0 0x0 0x0 0x2a 0x68
Release: 0xb1 0x0 0x11 0x6 0x50 0x0 0x14 0xa 0x5 0x0 0x4 0x0 0x0 0x1 0x7d 0x7a 0x37

I did some test code in a test file that seems to work:

```
playpress = bytearray(b'\xb1\x00\x10\x06\x50\x00\x14\x0a\x05\x04\x00\x00\x00\x00\x2a\x68')
playreleased = bytearray(b'\xb1\x00\x11\x06\x50\x00\x14\x0a\x05\x00\x04\x00\x00\x01\x7d\x7a\x37')

while True:
        tosend = bytearray(b'\x94\x06\x50\x6a')
        ser.write(tosend)
        expect = bytearray(b'\xb1\x00\x06\x06\x50\x0d')
        resp = bytearray(ser.read(10000))
        if resp != expect:
                if resp == playpress:
                        breaktime = time.time() + 3
                        buttonpress = BTNLONGPLAY
                        while time.time() < breaktime:
                                tosend = bytearray(b'\x94\x06\x50\x6a')
                                ser.write(tosend)
                                expect = bytearray(b'\xb1\x00\x06\x06\x50\x0d')
                                resp = bytearray(ser.read(10000))
                                if resp == playreleased:
                                        buttonpress = BTNPLAY
                                        break
                                time.sleep(0.1)
                        if buttonpress == BTNPLAY:
                                print("short press")
                        if buttonpress == BTNLONGPLAY:
                                print("long press")

        time.sleep(0.1)
```

**Reply by wormstein** [reply]
I thoought too much debug line might confuse you. I was actually checking just the 9th byte which turns to 0 when released... 
I was in panic mode I just didn't know what else to try.,,,

I also think this part could be turned into a function as I saw it all over the code. Looks like a Ping to the controller...
``` python
        tosend = bytearray(b'\x94\x06\x50\x6a')
        ser.write(tosend)
        expect = bytearray(b'\xb1\x00\x06\x06\x50\x0d')
        resp = bytearray(ser.read(10000))
```

**Reply by EdNekebno** [reply]
I see, yup I missed that, my brain isn't in coding mode this evening!

There are actually two pings that go on in the code in general:

```
    tosend = bytearray(b'\x83\x06\x50\x59')
    ser.write(tosend)
    expect = bytearray(b'\x85\x00\x06\x06\x50\x61')
    resp = ser.read(10000)
    resp = bytearray(resp)
```

and

```
    tosend = bytearray(b'\x94\x06\x50\x6a')
    ser.write(tosend)
    expect = bytearray(b'\xb1\x00\x06\x06\x50\x0d')
    resp = ser.read(10000)
    resp = bytearray(resp)
```

I think the first pings for pieces and the second for buttons. But I'm not 100% sure. I just made that assumption at one point and it seems to work! Both of the expected responses have no actual data in them (if you decode them with the format mentioned in this discussion: https://github.com/EdNekebno/DGTCentaur/discussions/14 ).

**Comment by EdNekebno** [comment]
Might want to be careful with the long press on the play button though :)

**Reply by wormstein** [reply]
Ha ha. No more than 10 :)) Caught me several times....

**Comment by wormstein** [comment]
Did some more tests. It's not very reliable yet as it sometimes says long press althouth it's short. It's too late to bang my head.... but it's good we know this exists.
